fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingBase.java

```diff
@@ -0,0 +1,370 @@
+
+        return AsyncUtil.whileTrue(() -> cursor.onNext().thenCompose(result -> {
+            if (!result.hasNext()) {
+                // end of the cursor list
+                timerIncrement(timer, FDBStoreTimer.Counts.ONLINE_INDEX_BUILDER_RANGES_BY_COUNT);
+                lastResultSet.accept(result);
+                return AsyncUtil.READY_FALSE;
+            }
+
+            final FDBStoredRecord<Message> rec = getRecord.apply(result);
+            isEmpty.set(false);
+            timerIncrement(timer, FDBStoreTimer.Counts.ONLINE_INDEX_BUILDER_RECORDS_SCANNED);
+            recordsScannedCounter.incrementAndGet();
+            if (!common.recordTypes.contains(rec.getRecordType())) {
+                // This record is not our type, swipe left
+                return AsyncUtil.READY_TRUE;
+            }
+            // add this index to the transaction
+            if (isIdempotent) {
+                store.addRecordReadConflict(rec.getPrimaryKey());
+            }
+            timerIncrement(timer, FDBStoreTimer.Counts.ONLINE_INDEX_BUILDER_RECORDS_INDEXED);
+
+            final CompletableFuture<Void> updateMaintainer = updateMaintainerBuilder(syntheticPlan, rec, maintainer, store);
+
+            return updateMaintainer.thenCompose(vignore ->
+                    context.getApproximateTransactionSize().thenApply(size -> {
+                        if (size >= common.config.getMaxWriteLimitBytes()) {
+                            // the transaction becomes too big - stop iterating
+                            timerIncrement(timer, FDBStoreTimer.Counts.ONLINE_INDEX_BUILDER_RANGES_BY_SIZE);
+                            lastResultSet.accept(result);
+                            return false;
+                        }
+                        return true;
+                    }));
+
+        }), cursor.getExecutor());
+    }
+
+    @Nonnull
+    public CompletableFuture<Void> rebuildIndexAsync(@Nonnull FDBRecordStore store) {
+        Index index = common.getIndex();
+        Transaction tr = store.ensureContextActive();
+        store.clearIndexData(index);
+
+        // Clear the associated range set (done as part of clearIndexData above) and make it instead equal to
+        // the complete range. This isn't super necessary, but it is done
+        // to avoid (1) concurrent OnlineIndexBuilders doing more work and
+        // (2) to allow for write-only indexes to continue to do the right thing.
+        RangeSet rangeSet = new RangeSet(store.indexRangeSubspace(index));
+        CompletableFuture<Boolean> rangeFuture = rangeSet.insertRange(tr, null, null);
+        CompletableFuture<Void> buildFuture = rebuildIndexByEntityAsync(store);
+
+        return CompletableFuture.allOf(rangeFuture, buildFuture);
+    }
+
+    abstract CompletableFuture<Void> rebuildIndexByEntityAsync(FDBRecordStore store);
+
+    // test support
```

Comment:
You mean "This method is only used by tests. Actually traffic uses throttle.throttledRunAsync directly"?